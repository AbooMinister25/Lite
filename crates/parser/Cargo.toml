[package]
name = "parser"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lexer = { path = "../lexer" }
error = { path = "../error" }
span = { path = "../span" }
ariadne = "0.1.5"
serde = { version = "1.0.144", features = ["derive"] }

[dev-dependencies]
insta = { version = "1.19.1", features = ["yaml"] }
