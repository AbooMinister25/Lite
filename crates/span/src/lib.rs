//! Contains the `Span` struct and `Spanned` type for representing the
//! position of different items in Lite throughout the source code.

#![warn(clippy::pedantic, clippy::nursery)]

use std::fmt;
use std::ops::Range;

pub type Spanned<T> = (T, Span);

#[derive(Copy, Clone, Debug, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl From<Range<usize>> for Span {
    fn from(r: Range<usize>) -> Self {
        Self {
            start: r.start,
            end: r.end,
        }
    }
}

impl fmt::Display for Span {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}..{}", self.start, self.end)
    }
}

impl ariadne::Span for Span {
    type SourceId = ();

    fn source(&self) -> &Self::SourceId {
        &()
    }

    fn start(&self) -> usize {
        self.start
    }

    fn end(&self) -> usize {
        self.end
    }
}
