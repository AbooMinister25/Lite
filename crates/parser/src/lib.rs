//! This file contains the main interface to the parser. The parser takes a
//! string and outputs an Abstract Syntax Tree (AST). Lite's parser is implemented
//! as a Pratt parser, and allows for top-down operator precedence parsing.

#![warn(clippy::pedantic, clippy::nursery)]
#![allow(clippy::must_use_candidate)]
#![allow(clippy::unnecessary_wraps)]
#![allow(clippy::missing_const_for_fn)]

pub mod ast;
pub mod errors;
pub mod expression;
pub mod precedence;
pub mod statement;

use crate::ast::Statement;
use crate::errors::{ErrorKind, ParserError};
use ariadne::{Color, ColorGenerator, Fmt, Label, Report, ReportBuilder, ReportKind, Source};
use lexer::{tokens::TokenKind, Lexer};
use span::{Span, Spanned};

/// Parses a string into an Abstract Syntax Tree (AST)
///
/// # Examples
///
/// ```
/// use parser::Parser;
///
/// let source = r#"
/// func main() do
///     println("Hello World")
/// end
/// "#;
///
/// let mut parser = Parser::new(source, "main.lt")
/// ```
pub struct Parser<'a> {
    source: &'a str,
    lexer: Lexer<'a>,
    filename: &'a str,
    current_token_span: Span,
    peeked: Option<Spanned<TokenKind>>,
}

impl<'a> Parser<'a> {
    /// Constructs a new `Parser`.
    ///
    /// # Examples
    ///
    /// ```
    /// use parser::Parser;
    ///
    /// let source = r#"
    /// func main() do
    ///     println("Hello World")
    /// end
    /// "#;
    ///
    /// let mut parser = Parser::new(source, "main.lt");
    /// ```
    pub fn new(source: &'a str, filename: &'a str) -> Self {
        Self {
            source,
            lexer: Lexer::new(source),
            filename,
            current_token_span: Span::from(0..0),
            peeked: None,
        }
    }

    /// Parses and returns a tuple of the parsed AST and any errors
    /// that may have occurred.
    ///
    /// # Examples
    ///
    /// ```
    /// use parser::Parser;
    ///
    /// let source = r#"
    /// func main() do
    ///     println("Hello World")
    /// end
    /// "#;
    ///
    /// let mut parser = Parser::new(source, "main.lt");
    /// let (ast, errors) = parser.parse();
    /// ```
    pub fn parse(&mut self) -> (Vec<Spanned<Statement>>, Vec<ParserError>) {
        let mut nodes = vec![];
        let mut errors = vec![];

        while !self.at_end() {
            let node = self.parse_statement();
            match node {
                Ok(n) => nodes.push(n),
                Err(e) => errors.push(e),
            };
        }

        (nodes, errors)
    }

    fn at_end(&mut self) -> bool {
        self.peek().0 == TokenKind::EoF
    }

    fn advance(&mut self) -> Spanned<TokenKind> {
        // If a token has been peeked, return that, otherwise advance the lexer and return the next token
        if let Some(t) = self.peeked.take() {
            self.current_token_span = t.1;
            t
        } else {
            let t = self.lexer.next_token();
            self.current_token_span = t.1;
            t
        }
    }

    fn peek(&mut self) -> &Spanned<TokenKind> {
        // If nothing has been peeked, advance and store that token as the peeked value
        if self.peeked.is_none() {
            self.peeked = Some(self.advance());
        }

        self.peeked.as_ref().unwrap()
    }

    fn consume(&mut self, expected: &TokenKind, message: &str) -> Result<(), ParserError> {
        let token = self.peek();

        if token.0 == *expected {
            self.advance(); //  next token was the expected one, so advance
            return Ok(());
        }

        Err(ParserError::new(
            ErrorKind::Expected(vec![expected.to_string()], token.0.clone(), token.1),
            message.to_string(),
            None,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ast::{Expr, LiteralKind};

    #[test]
    fn its_alive() {
        let mut parser = Parser::new("5", "main.lt");
        let (ast, errors) = parser.parse();

        assert!(errors.is_empty());
        assert_eq!(
            ast,
            vec![(
                Statement::Expression((Expr::Literal(LiteralKind::Int(5)), Span::from(0..1))),
                Span::from(0..1)
            )]
        );
    }
}
