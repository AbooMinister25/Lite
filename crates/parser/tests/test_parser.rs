use parser::{ast::Statement, Parser};
use span::Spanned;

fn parse_string(string: &str) -> Vec<Spanned<Statement>> {
    let mut parser = Parser::new(string, "test.lt");
    let (ast, _) = parser.parse();

    ast
}

#[test]
fn parse_literals() {
    let test_str = r#"10 3.5 true false "hello world" 'c'"#;
    let ast = parse_string(test_str);

    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_idents() {
    let test_str = "foo _bar a123";
    let ast = parse_string(test_str);

    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_tuple() {
    let test_str = r"(1, 2, 3)";
    let ast = parse_string(test_str);

    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_array() {
    let test_str = "[1, 2, 3]";
    let ast = parse_string(test_str);

    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_unary() {
    let test_str = "!foo";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);

    let test_str = "-bar";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_binary() {
    let test_str = "5 + 5 - 3 / 2";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_call() {
    let test_str = "foo(a, b, c)";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_assignment() {
    let test_str = "a = 10";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_block() {
    let test_str = "
do
    println(10)
end
    ";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_if() {
    let test_str = r#"
if 1 == 1 do
    println("All is right with the world")
else if 1 == 2 do
    println("no its not??")
else do
    println("what the heck")
end
    "#;
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_for() {
    let test_str = "
for i in [1, 2, 3] do
    println(i)
end
    ";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_while() {
    let test_str = r#"
while 1 == 1 do
    println("All is right with the world")
end
    "#;
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_return() {
    let test_str = "return 10";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_import() {
    let test_str = "import foo";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_let() {
    let test_str = "
let foo = 10
let mut foo = 20
    ";
    let ast = parse_string(test_str);
    insta::assert_yaml_snapshot!(ast);
}

#[test]
fn parse_function() {
    let test_str = r#"
func foo(a: int, b: int) do
    return 10
end
    "#;
    let ast = parse_string(test_str);

    insta::assert_yaml_snapshot!(ast);
}
