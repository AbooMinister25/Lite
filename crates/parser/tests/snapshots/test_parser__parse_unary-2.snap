---
source: crates/parser/tests/test_parser.rs
expression: ast
---
- - Expression:
      - Unary:
          op: "-"
          rhs:
            - Ident: bar
            - start: 1
              end: 4
      - start: 0
        end: 4
  - start: 0
    end: 4

